### PR TITLE
Add configurable number of IO dispatchers and synchronize on MongoDocument.putAll

### LINKED ISSUE #349: Race condition when updating documents when multiple stages are handling documents at the same time
Using the SetStaticActionStage and at the same time running other stages, the status is not always updated.

Even though the document is fetched and touched by the SetStaticActionStage, the action is not changed.

### PR DESCRIPTION
**Pull request by Nycander**
Fixes #349

### CODE REVIEW

File: core/src/main/java/com/findwise/hydra/net/WriteHandler.java [hunk 1]
```diff
@@ -118,21 +118,19 @@ private boolean handlePartialWrite(DatabaseDocument<T> md, HttpResponse response
...
+		    HttpResponseWriter.printNoDocument(response);
+		    return false;
+		}
 
-		if(io.update(inDB)){
-			HttpResponseWriter.printSaveOk(response, md.getID());
-			return true;
-		} 
+		if(io.update(md)){
+		    HttpResponseWriter.printSaveOk(response, md.getID());
```
Comment (ssimon) [inline core/src/main/java/com/findwise/hydra/net/WriteHandler.java]: Formatting

File: database-impl/mongodb/src/main/java/com/findwise/hydra/mongodb/MongoDocument.java [hunk 2]
```diff
@@ -239,22 +239,21 @@ else if(d.getID() instanceof MongoDocumentID) {
...
 				return;
 			}
 		}
-		
-		
-		if(d.getAction()!=null) {
-			setAction(d.getAction());
+
+		if (d.getAction() != null) {
+		    setAction(d.getAction());
```
Comment (ssimon) [inline database-impl/mongodb/src/main/java/com/findwise/hydra/mongodb/MongoDocument.java]: Formatting

File: test-suite/functional-tests/src/test/java/com/findwise/hydra/ParallelPipelineBuilder.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ParallelPipelineBuilder {
+    private String stageGroupName;
```
Comment (ssimon) [inline test-suite/functional-tests/src/test/java/com/findwise/hydra/ParallelPipelineBuilder.java]: Formatting

File: test-suite/functional-tests/src/test/java/com/findwise/hydra/RaceConditionIT.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+import java.util.UUID;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+public class RaceConditionIT {
+
+    private MongoConfiguration mongoConfiguration;
+    private MongoConnector mongoConnector;
```
Comment (ssimon) [inline test-suite/functional-tests/src/test/java/com/findwise/hydra/RaceConditionIT.java]: Formatting

File: test-suite/functional-tests/src/test/resources/logback.xml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<configuration>
+
+  <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
```
Comment (ssimon) [inline test-suite/functional-tests/src/test/resources/logback.xml]: Formatting

File: test-suite/integration-test-stages/src/main/java/com/findwise/hydra/stage/SleepyStage.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
+package com.findwise.hydra.stage;
+
+import com.findwise.hydra.local.LocalDocument;
+
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+@Stage
+public class SleepyStage extends AbstractProcessStage {
```
Comment (ssimon) [inline test-suite/integration-test-stages/src/main/java/com/findwise/hydra/stage/SleepyStage.java:10]: Formatting

File: core/src/main/java/com/findwise/hydra/CoreMapConfiguration.java [hunk 7]
```diff
@@ -5,7 +5,7 @@
 
 public class CoreMapConfiguration implements CoreConfiguration, Configuration {
 
-	private final MongoConfiguration databaseConfiguration;
+    private final MongoConfiguration databaseConfiguration;
```
Comment (ssimon) [inline core/src/main/java/com/findwise/hydra/CoreMapConfiguration.java]: Formatting

File: test-suite/integration-test-stages/src/main/java/com/findwise/hydra/stage/SleepyStage.java (hunk 6 above)
Comment (Nycander) [inline test-suite/integration-test-stages/src/main/java/com/findwise/hydra/stage/SleepyStage.java:10]: You sure? The file is all formatted the same way.
Do we have a code style defined for Hydra? ping @laserval

Comment (laserval) [inline test-suite/integration-test-stages/src/main/java/com/findwise/hydra/stage/SleepyStage.java:10]: Current code style in Hydra is to use tabs. Go with tabs for now.

We might want to move to spaces, but for now I'd prioritize consistency.

**Comment by laserval** [comment]
One test in database is failing: `testGetAndTagCacheMiss()`. It uses tons of mocks, but it seems that it is assuming that the object returned by the backing database is the same object that is returned by the cache. Is this not true anymore?

File: test-suite/integration-test-stages/src/main/java/com/findwise/hydra/stage/SleepyStage.java (hunk 6 above)
Comment (Nycander) [inline test-suite/integration-test-stages/src/main/java/com/findwise/hydra/stage/SleepyStage.java:10]: Fixed

**Comment by Nycander** [comment]
@laserval The test merely tests that someone has executed setFetchedBy on the document, in the case of a cache miss.

This is done by both Database implementations, this does not show since the database implementation is mocked away. I vote removing the assertion.
A commit is coming up for that.

**Comment by laserval** [comment]
This is failing on the new integration test on Travis, see:
https://s3.amazonaws.com/archive.travis-ci.org/jobs/41457246/log.txt

I also had some problems running it, the test VM exited. Will have to look into that - might be the instability of this. multiprocess stuff...

**Comment by Nycander** [comment]
Yes. We might have to remove the test, it isn't stable at all.
What do you think?

Perhaps keeping the test, but disabling it?

**Pull request merged**